### PR TITLE
feat: create basic functionality for Extensions

### PR DESCRIPTION
**Pull request by Catalyst4222**
Expands Extension.__new__ and adds new functions to mimic client decorators

## About

This pr adds in basic functionality for the Extension class, along with decorators for commands/components/listeners
Marked as a draft because some things are missing, like properly removing commands on removal of the Extension
Any suggestions on improvements are welcome

## Checklist

- [ ] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [x] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent): [[MISC] Revamping the cogs/extension framework](https://github.com/goverfl0w/discord-interactions/issues/425)
- [x] I've made this pull request for/as: (check all that apply)
  - [ ] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

**Comment by Catalyst4222** [comment]
I would like someone to double check that everything works, and I'm not just hallucinating a lack of errors. It would probably be a good idea to simplify some of the code, I got it working but it doesn't look the nicest

**Comment by i0bs** [comment]
Delta and I will test this code when this PR is marked ready for reviewing

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
from functools import wraps
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
@wraps(command)
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
@wraps(Client.component)
```

Please either uncomment this or remove it if you do not plan on using it.

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
from inspect import getmembers
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
            for ext_name, ext in getmembers(
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
        for name, func in getmembers(self, predicate=iscoroutinefunction):
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
from types import ModuleType
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
        if isinstance(extension, ModuleType):
```

File: interactions/client.pyi
Comment (i0bs) [inline interactions/client.pyi]:
```suggestion
```
This isn't used at all anywhere else in the script (from where I can tell)

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]: `["_".join(cmd.split("_")[1:])`

This is a messy operation that I highly dislike. Why is this being done? Let's say we have a command named as `hello-world`, this won't matter any. But if it uses snake_case like `long_function_name` then you're running the same thing as without the wildcard. You could just do `cmd[7:]` since you're encompassing the length of `command_` here.

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
            lambda cmd: cmd["name"] in clean_cmd_names,
```

Try and step away from alphabet soup abstraction, `x` may make sense from a maintainer's perspective but anyone else reading this can easily get confused by common iterator naming conventions

**Review by i0bs — CHANGES_REQUESTED** [review]

**Comment by FayeDel** [comment]
There's support for `@client.component`, `@Client.command`, etc using these, but what about `@Client.autocomplete`?

**Comment by i0bs** [comment]
No support for modals either

File: interactions/client.py
Comment (mAxYoLo01) [inline interactions/client.py]:
```suggestion
        return self.load(name, package, *args, **kwargs)
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
        :rtype: Optional[Extension]
```

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py]:
```suggestion
        :rtype: Optional[Extension]
```

**Review by i0bs — CHANGES_REQUESTED** [review]

**Review by i0bs — APPROVED** [review]
LGTM. Please pre-commit though.

File: interactions/client.py
Comment (mAxYoLo01) [inline interactions/client.py]:
```suggestion
        client._extensions[cls.__name__] = self
```
With this small fix I was able to create a working basic cog command

**Comment by i0bs** [comment]
This is LGTM, I think we're finally ready to go ahead and merge this. I'll be leaving the last approval on this for @DeltaXWizard , but all of these changes look good on my end.

**Comment by FayeDel** [comment]
You may need to run pre-commit again after new pending changes, I'll test afterwards though :)
(Code looks good, I have no objections, I just need to test the thing before I approve)

File: interactions/client.py
Comment (mAxYoLo01) [inline interactions/client.py:654]:
```suggestion
            log.error(f"Extension {name} has already been loaded. Skipping.")
            return
```

**Comment by mAxYoLo01** [comment]
After some more tests, I figured out that there still are a few problems to address:
1. When loading any extension **after** `bot.start()`, all commands that are initialized outside of a cog (for example initialized in the main launch file) will all disappear from the Slash Commands Menu. Commands of the loaded cog will not be removed from the Menu but will not respond either.
2. After unloading an extension, the commands from this extension are not removed, and can still be accessed from the Slash Commands Menu.

Feel free to ask me if you need some example code to recreate these issues.

**Comment by Toricane** [comment]
I don't think you are supposed to load extensions after `bot.start()`
But it does seem like an issue, shouldn't be wiping the ones in the main file

**Comment by Catalyst4222** [comment]
I want to test again, just to make sure I fixed the breaking changes from the last patch

**Comment by mAxYoLo01** [comment]
> I don't think you are supposed to load extensions after `bot.start()` But it does seem like an issue, shouldn't be wiping the ones in the main file

Didn't know it might be a possibility that it shouldn't be loaded afterwards, in fact I was creating a Slash Command that allows the user to un/re/load cogs directly via Discord. I've previously done it with d.py, so I thought that it would be a feature that i.py should also have.

**Comment by sillyElvy** [comment]
Afaik fl0w wanted something like hot reload for cogs at first so it should be possible tho shouldn't it?

**Comment by Toricane** [comment]
> Afaik fl0w wanted something like hot reload for cogs at first so it should be possible tho shouldn't it?

Oh ok never mind then

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py:656]: For what it's worth, apparently this is no longer deleting the actual module from the system. So, this is why your extensions are still working after they've been removed. Deleting it from the list of extensions stored in the Client is not a viable thing.

File: interactions/client.py
Comment (i0bs) [inline interactions/client.py:706]:
```suggestion
        del sys.modules[_name]
        del self._extensions[_name]
```

**Review by i0bs — CHANGES_REQUESTED** [review]

File: interactions/client.py
Comment (Catalyst4222) [reply interactions/client.py:706]: That causes an error when unloading something that isn't a module, e.x. an Extension

**Comment by Catalyst4222** [comment]
I don't know how to do pre-commit
I do `git add <changes>`, `black.exe .`, `isort.exe .`, `flake8.exe .` and do the changes, but it *still* failes

**Comment by Catalyst4222** [comment]
image

**Comment by mAxYoLo01** [comment]
Alright there is still the problem that "global" commands (not initialized in cogs) disappear after loading any cog on my side. The other issues have been fixed though.

**Comment by Catalyst4222** [comment]
> "global" commands (not initialized in cogs) disappear after loading any cog on my side

I'm unable to reproduce this. All commands outside of the cog remain usable, and any inside get added/removed as the cog gets loaded/unloaded. The only reason I can think of for it happening would be global commands, as I've only been testing in the guild scope. I'll try global commands now, but I don't expect a difference

**Comment by mAxYoLo01** [comment]
> The only reason I can think of for it happening would be global commands, as I've only been testing in the guild scope. 

Oh my bad on this one, I formulated badly my previous message. In fact, I also have tried it only with a guild scope. 
Weird that we don't have the same behaviour, on my hand whenever I load/reload a cog, all the commands outside of this cog disappear.

**Comment by Catalyst4222** [comment]
I've recreated the missing commands issue, and it seems to only happen when a cog is unloaded then reloaded in quick succession. Any ideas as to why or where this is happening is welcome

**Comment by i0bs** [comment]
I'll be moving through with merging this to `unstable`, I'm looking to hit a feature lock on getting extensions in rather soon and we can always make revisions with the pull request being brought in. For any relating concerns to this, I advise others to create an [Issue](https://github.com/interactions-py/library/issues) or make a mention of it in the #unstable-dev channel in our support server.

**Pull request merged**
